File: Tyche.Android/bet/src/main/java/com/felipearpa/tyche/bet/pending/PendingBetItemViewModel.kt
package com.felipearpa.tyche.bet.pending

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.felipearpa.tyche.data.bet.application.BetUseCase
import com.felipearpa.tyche.data.bet.domain.Bet
import com.felipearpa.tyche.data.bet.domain.BetException
import com.felipearpa.tyche.bet.PoolGamblerBetModel
import com.felipearpa.tyche.bet.toPoolGamblerBetModel
import com.felipearpa.tyche.core.type.BetScore
import com.felipearpa.tyche.core.type.TeamScore
import com.felipearpa.tyche.ui.exception.orDefaultLocalized
import com.felipearpa.ui.state.EditableViewState
import com.felipearpa.ui.state.relevantValue
import dagger.assisted.Assisted
import dagger.assisted.AssistedInject
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.flow.updateAndGet
import kotlinx.coroutines.launch
import java.util.UUID

class PendingBetItemViewModel @AssistedInject constructor(
    @Assisted poolGamblerBet: PoolGamblerBetModel,
    private val betUseCase: BetUseCase
) :
    ViewModel() {

    private val _state =
        MutableStateFlow<EditableViewState<PoolGamblerBetModel>>(
            EditableViewState.Initial(poolGamblerBet)
        )
    val state = _state.asStateFlow()

    fun retryBet() {
        val poolGamblerBet = when (val stateValue = this.state.value) {
            is EditableViewState.Failure -> stateValue.failed
            else -> return
        }
        val betScore = poolGamblerBet.betScore ?: return
        bet(betScore = betScore)
    }

    fun bet(betScore: TeamScore<Int>) {
        viewModelScope.launch {
            val (currentPoolGamblerBet, targetPoolGamblerBet) = _state.updateAndGet { currentState ->
                val currentPoolGamblerBet = currentState.relevantValue()
                val targetPoolGamblerBet = currentPoolGamblerBet.copy(betScore = betScore)
                EditableViewState.Loading(
                    current = currentPoolGamblerBet,
                    target = targetPoolGamblerBet
                )
            } as EditableViewState.Loading

            val bet = Bet(
                poolId = currentPoolGamblerBet.poolId,
                gamblerId = currentPoolGamblerBet.gamblerId,
                matchId = currentPoolGamblerBet.matchId,
                homeTeamBet = BetScore(betScore.homeTeamValue),
                awayTeamBet = BetScore(betScore.awayTeamValue)
            )

            betUseCase.execute(bet = bet)
                .onSuccess { updatedPoolGamblerBet ->
                    _state.emit(
                        EditableViewState.Success(
                            old = currentPoolGamblerBet,
                            succeeded = updatedPoolGamblerBet.toPoolGamblerBetModel()
                        )
                    )
                }.onFailure { exception ->
                    if (exception is BetException.Forbidden) {
                        _state.emit(EditableViewState.Initial(currentPoolGamblerBet.copy(isLocked = true)))
                    } else {
                        _state.emit(
                            EditableViewState.Failure(
                                current = currentPoolGamblerBet,
                                failed = targetPoolGamblerBet,
                                exception = exception
                                    .toBetLocalizedExceptionOnMatch()
                                    .orDefaultLocalized()
                            )
                        )
                    }
                }
        }
    }

    fun reset() {
        _state.update { currentState ->
            val poolGamblerBet = currentState.relevantValue()
            EditableViewState.Initial(
                poolGamblerBet.copy(
                    instanceId = UUID.randomUUID().toString()
                )
            )
        }
    }
}

File: iOS/Bet/Sources/Bet/Pending/PendingBetItemViewModel.swift
import Foundation
import UI
import Core
import DataBet
import ViewingState

class PendingBetItemViewModel: ObservableObject {
    private let betUseCase: BetUseCase

    @Published @MainActor private(set) var state: MutationState<PoolGamblerBetModel>?

    @MainActor
    public init(betUseCase: BetUseCase) {
        self.betUseCase = betUseCase
        self.state = nil
    }

    @MainActor
    func bind(_ poolGamblerBet: PoolGamblerBetModel) {
        switch state {
        case nil, .idle:
            state = .idle(poolGamblerBet)
        case .mutated(let old, _):
            state = .mutated(original: old, updated: poolGamblerBet)
        case .mutating, .failure:
            break
        }
    }

    @MainActor
    func retryBet() {
        guard case .failure(_, updated: let poolGamblerBet, _) = state else { return }
        guard let betScore = poolGamblerBet.betScore else { return }
        bet(betScore: betScore)
    }

    @MainActor
    func bet(betScore: TeamScore<Int>) {
        Task { [self] in
            guard let currentState = state else { return }
            var currentPoolGamblerBet = currentState.activeValue()

            let targetPoolGamblerBet = currentPoolGamblerBet.copy { builder in
                builder.betScore = betScore
            }

            state = .mutating(original: currentPoolGamblerBet, updated: targetPoolGamblerBet)

            let bet = Bet(
                poolId: targetPoolGamblerBet.poolId,
                gamblerId: targetPoolGamblerBet.gamblerId,
                matchId: targetPoolGamblerBet.matchId,
                homeTeamBet: BetScore(betScore.homeTeamValue)!,
                awayTeamBet: BetScore(betScore.awayTeamValue)!
            )
            let result = await betUseCase.execute(bet: bet)

            switch result {
            case .success(let updatedPoolGamblerBet):
                state = .mutated(
                    original: currentPoolGamblerBet,
                    updated: updatedPoolGamblerBet.toPoolGamblerBetModel()
                )
            case .failure(let error):
                if case BetError.forbidden = error {
                    currentPoolGamblerBet.lock()
                    state = .idle(currentPoolGamblerBet)
                } else {
                    state = .failure(
                        original: currentPoolGamblerBet,
                        updated: targetPoolGamblerBet,
                        error: error.mapOrDefaultLocalized { error in error.asBetLocalizedError() }
                    )
                }
            }
        }
    }

    @MainActor
    func reset() {
        guard let currentState = state else { return }
        state = .idle(currentState.activeValue())
    }
}
